berkdb/mp/mp_sync.c

```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
+	hash_t *fileid_pages = NULL;
+
+	dbmp = dbenv->mp_handle;
+	mp = dbmp->reginfo[0].primary;
+
+	(*pages) = 0;
+	fileid_pages = hash_init(DB_FILE_ID_LEN);
+	for (n_cache = 0; n_cache < mp->nreg; ++n_cache) {
+		c_mp = dbmp->reginfo[n_cache].primary;
+
+		hp = R_ADDR(&dbmp->reginfo[n_cache], c_mp->htab);
+		for (i = 0; i < c_mp->htab_buckets; i++, hp++) {
+
+			if (SH_TAILQ_FIRST(&hp->hash_bucket, __bh) == NULL)
+				continue;
+
+			MUTEX_LOCK(dbenv, &hp->hash_mutex);
+			for (bhp = SH_TAILQ_FIRST(&hp->hash_bucket, __bh);
+				bhp != NULL; bhp = SH_TAILQ_NEXT(bhp, hq, __bh)) {
+
+				mfp = R_ADDR(dbmp->reginfo, bhp->mf_offset);
+
+				if (F_ISSET(mfp, MP_TEMP) || mfp->lsn_off == -1)
+					continue;
+
+				mfp = R_ADDR(dbmp->reginfo, bhp->mf_offset);
+				fileid = R_ADDR(dbmp->reginfo, mfp->fileid_off);
+
+				if ((ret = add_fileid_page(dbenv, fileid_pages, fileid,
+								bhp->pgno)) != 0) {
+					MUTEX_UNLOCK(dbenv, &hp->hash_mutex);
+					destroy_fileid_page_hash(dbenv, fileid_pages);
+					logmsg(LOGMSG_ERROR, "%s error adding fileid page to hash "
+							"%d\n", __func__, ret);
+					return ret;
+				}
+				(*pages)++;
+			}
+			MUTEX_UNLOCK(dbenv, &hp->hash_mutex);
+		}
+	}
+
+    output_fileid_page_hash(dbenv, fileid_pages, s);
+    destroy_fileid_page_hash(dbenv, fileid_pages);
+
+	return 0;
+}
+
+static pthread_mutex_t page_flush_lk = PTHREAD_MUTEX_INITIALIZER;
+
+#define PAGELIST "pagelist"
+#define PAGELISTTEMP "pagelist.tmp"
+
+char *bdb_trans(const char infile[], char outfile[]);
+
+/*
+ * __memp_load_pagelist --
+ *	Load bufferpool pagelist to log directory
+ *
+ * PUBLIC: int __memp_load_pagelist
+ * PUBLIC:	 __P((DB_ENV *));
+ */
+int
+__memp_load_pagelist(dbenv)
+	DB_ENV *dbenv;
+{
+	char path[PATH_MAX], pathbuf[PATH_MAX], *rpath;
+	u_int32_t cnt, lines;
+	int fd, ret;
+	SBUF2 *s;
+
+	Pthread_mutex_lock(&page_flush_lk);
+
+	snprintf(path, sizeof(path), "%s/%s", dbenv->db_home, PAGELIST);
+	rpath = bdb_trans(path, pathbuf);
+#if PAGELIST_DEBUG
+	logmsg(LOGMSG_USER, "%s line %d opening %s\n", __func__, __LINE__,
+			rpath);
+#endif
+	if ((fd = open(rpath, O_RDONLY, 0666)) < 0 ||
+			(s = sbuf2open(fd, 0)) == NULL) {
+#if PAGELIST_DEBUG
+		logmsg(LOGMSG_ERROR, "%s line %d error opening %s, %d\n", __func__,
+				__LINE__, rpath, errno);
+#endif
+		if (fd >= 0)
+			close(fd);
+		ret = -1;
+		goto done;
+	}
+
+	if ((ret = __memp_load(dbenv, s, &cnt, &lines)) != 0) {
+#if PAGELIST_DEBUG
+		logmsg(LOGMSG_USER, "%s failed load page cache: bufferpool not open\n",
+				__func__);
+#endif
+	} else {
+#if PAGELIST_DEBUG
+		logmsg(LOGMSG_USER, "%s loaded cache %u pages processed %u lines, "
+				"ret=%d\n", __func__, cnt, lines, ret);
+#endif
+	}
+	sbuf2close(s);
+
+done:
+	Pthread_mutex_unlock(&page_flush_lk);
+	return ret;
+}
+
+/*
+ * __memp_flush_pagelist --
+ *	Dump bufferpool pagelist to log directory
+ *
+ * PUBLIC: int __memp_flush_pagelist
+ * PUBLIC:	 __P((DB_ENV *));
+ */
+int
```

Comment:
Introduce a tunable that will short-circuit this logic unless a percentage of the bufferpool has change (we can detect misses, and therefore the number of memp_fgets that result in IO, and modify the bufferpool)